### PR TITLE
Improve error handling for ElfFile

### PR DESCRIPTION
**Pull request by dimitry-**
Split construction and initialization of ElfFile,
report and propogate errors that happen on initialization.

Test: run ObjectFileTests

### CODE REVIEW

File: src/ObjectUtils/ElfFile.cpp [hunk 1]
```diff
@@ -563,11 +602,15 @@ ErrorMessageOr<std::unique_ptr<ElfFile>> CreateElfFile(
...
+        std::make_unique<ElfFileImpl<llvm::object::ELF32LE>>(file_path, std::move(file));
+    OUTCOME_TRY(instance->Initialize());
+    result = std::move(instance);
   } else if (llvm::dyn_cast<llvm::object::ELF64LEObjectFile>(object_file) != nullptr) {
-    result = std::unique_ptr<ElfFile>(
-        new ElfFileImpl<llvm::object::ELF64LE>(file_path, std::move(file)));
+    auto instance =
+        std::make_unique<ElfFileImpl<llvm::object::ELF64LE>>(file_path, std::move(file));
+    OUTCOME_TRY(instance->Initialize());
+    result = std::move(instance);
```
Comment (beckerhe) [inline src/ObjectUtils/ElfFile.cpp:613]: I assume they code was different at some point, but now I don't see a reason for having `result`. Why not return early in each of these cases?

**Review by beckerhe — APPROVED** [review]
Nice. Otherwise LGTM

File: src/ObjectUtils/ElfFile.cpp (hunk 1 above)
Comment (dimitry-) [reply src/ObjectUtils/ElfFile.cpp:613]: makes sense. done.

File: src/ObjectUtils/ElfFile.cpp [hunk 2]
```diff
@@ -178,57 +197,66 @@ void ElfFileImpl<ElfT>::InitDynamicEntries(const llvm::object::ELFFile<ElfT>* el
...
   }
 
   auto strtab_last_byte_or_error = elf_file->toMappedAddr(dynamic_string_table_addr.value() +
                                                           dynamic_string_table_size.value() - 1);
   if (!strtab_last_byte_or_error) {
-    LOG("Unable get last byte address of dynamic string table \"%s\": %s", file_path_.string(),
-        llvm::toString(strtab_last_byte_or_error.takeError()));
-    return;
+    auto error_message =
+        absl::StrFormat("Unable get last byte address of dynamic string table \"%s\": %s",
```
Comment (ronaldfw) [inline src/ObjectUtils/ElfFile.cpp:207]:
```suggestion
        absl::StrFormat("Unable to get last byte address of dynamic string table \"%s\": %s",
```

File: src/ObjectUtils/include/ObjectUtils/ElfFile.h [hunk 3]
```diff
@@ -53,7 +53,8 @@ class ElfFile : public ObjectFile {
   //
   // This method returns load bias for the elf-file if program headers are
   // available. This should be the case for all loadable elf-files.
-  [[nodiscard]] virtual ErrorMessageOr<uint64_t> GetLoadBias() const = 0;
+  [[nodiscard]] virtual uint64_t GetLoadBias() const = 0;
+  [[nodiscard]] virtual uint64_t GetExecutableSegmentOffset() const = 0;
```
Comment (ronaldfw) [inline src/ObjectUtils/include/ObjectUtils/ElfFile.h:57]: Is this a function you need for the absolute address changes you are working on? (Just wondering, as this is not yet used anywhere other than tests it seems.)

**Review by ronaldfw — APPROVED** [review]
LGTM

I suppose similar changes can be / need to be done for the COFFFile class eventually? (I can take care of these later today, if we need them. I need to look at the code there anyway.)

File: src/ObjectUtils/include/ObjectUtils/ElfFile.h (hunk 3 above)
Comment (dimitry-) [reply src/ObjectUtils/include/ObjectUtils/ElfFile.h:57]: Yes, it is going to be used in the next PR; the initial change was too big - so I split it to 2.

File: src/ObjectUtils/ElfFile.cpp (hunk 2 above)
Comment (dimitry-) [reply src/ObjectUtils/ElfFile.cpp:207]: done.

Comment (dimitry-) [reply src/ObjectUtils/ElfFile.cpp]: done.

**Pull request merged**